### PR TITLE
DB hierarchy refactor

### PR DESCRIPTION
**Pull request by CatalinVoss**
These changes refactor in our model hierarchy to eliminate `Context` as the in-between uber-class that everything inherits from. Rather than having corresponding entries in the `context` table and operating via joined table inheritance, `Sentence`, `Document`, `Span`, etc. all now have their own tables. This speeds up insertion and other queries significantly.

This means that `candidate_subclass()` now needs to keep track of the types of its args.

@ajratner This is somewhat invasive and related to our discussion. I find the code in `PretaggedCandidateExtractorUDF` pretty hard to follow, but this does pass all of our tests for labeling & generative model on our side now.

### CODE REVIEW

**Comment by ajratner** [comment]
Okay looked through- I totally get the motivation here, but definitely a big change :) . One route is that this is a separate "flat" DB structure so the user could choose this instead of the hierarchical one... will think about more tomorrow!

**Comment by CatalinVoss** [comment]
OK, let's discuss on Monday. My guess is that most users may not want to choose this. I'll run some benchmarks on our end to quantify the performance gain a bit in the meantime.

**Comment by CatalinVoss** [comment]
@ajratner Any thoughts on https://github.com/HazyResearch/snorkel/pull/1031/commits/32c5dd66e36d3b266f9e2170d79eb6278e0e1d7f ? Seems like this should ideally be a bit more asynchronous…

**Comment by henryre** [comment]
Hi @CatalinVoss, we've moved the information extraction version of Snorkel (v0.7) to https://github.com/snorkel-team/snorkel-extraction. Feel free to reopen this PR there!
